### DISCUSSION TITLE
what is the best option for machine learning on esp32(s3)?

### DISCUSSION THREAD

**Discussion by kaeoticEve** [micropython/micropython#10786] [General]
Good day,
I'm trying to detect spikes and categorize patterns from the i2s audio signals with the[ESP32-S3-DevKitC-1](n16r8), similar to micro speech. I figure that I need to have ulab for FFT and some machine-learning framework to build a categorizer.
I was able to compile the ulab into the latest firmware as a user module, but for the ML frameworks, I've attempted with the tflm examples (https://github.com/mocleiri/tensorflow-micropython-examples), the build process returns success but all the s3 firmware doesn't run on my device (including the ones that I build on my computer and the ones from GitHub action artifacts). after flash the device will be stuck in a boot loop and prints out esp back trance in the serial port.
I've also attempted to build with the upytorch(https://github.com/ljk53/upytorch), but I have not had any success so far.
I find that all the projects above are using the CMAKE differently, the ulab uses the standard user module mechanism provided by micropython, but as someone with little C language experience, I find it quite difficult to understand the build mechanism the other two uses.
which framework should I use? or should I build a minimal neural net runner that accepts trained weights?
thanks,

**Comment by jonnor** [comment]
Hi @JiewenGuan . It seems that not so many are working on Machine Learning with MicroPython (yet!). Did you get closer to something working?

**Comment by jonnor** [comment]
I have started https://github.com/emlearn/emlearn-micropython - an implementation of scikit-learn compatible models for MicroPython.
Having simple installation and deployment is a priority in the project. Right now, it only provides a RandomForest classifier, so it is a while away for sound event detection.

But emlearn has mel-spectrogram support, so we could add MicroPython bindings for that.
Combined with the existing classifiers, this should be good enough for basic Sound Event Detection.
More advanced events are probably best done with convolutionral and/or recurrent neural networks.

**Comment by sophiedegran** [comment]
Jiewen I found also that can be interest for you (I hope).
See github.com/atomic14/tensorflow-micropython-examples
Tell me!
